core/domain/prod_validation_jobs_one_off.py

```diff
@@ -75,58 +66,153 @@ def _validate_external_id_relationships(cls, item):
+                collection_models.CollectionModel,
+                item.collection_ids),
             'general_feedback_thread_ids': (
-                feedback_models.GeneralFeedbackThreadModel),
-            'creator_ids': user_models.UserSettingsModel,
-            'id': user_models.UserSettingsModel,
+                feedback_models.GeneralFeedbackThreadModel,
+                item.general_feedback_thread_ids),
+            'creator_ids': (user_models.UserSettingsModel, item.creator_ids),
+            'id': (user_models.UserSettingsModel, [item.id]),
         }
 
     @classmethod
-    def validate(cls, item):
-        """Validates a UserSubscriptionsModel entity.
+    def _get_validation_functions(cls):
+        return []
+
+
+class ExplorationModelValidator(BaseModelValidator):
+    """Class for validating ExplorationModel."""
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        state_id_mapping_model_ids = [
+            '%s.%d' % (item.id, version) for version in range(
+                1, item.version + 1)]
+        return {
+            'state_id_mapping_model': (
+                exp_models.StateIdMappingModel,
+                state_id_mapping_model_ids)
+        }
+
+    @classmethod
+    def _validate_state_name(cls, item):
+        """Validate that state name of StateIdMappingModel matches
+        corresponding ExplorationModel states.
 
         Args:
-            item: UserSubscriptionsModel. Entity to validate.
+            item: ndb.Model. Entity to validate.
         """
-        cls.errors.clear()
+        _, state_id_mapping_model_tuples = (
+            cls.external_models['state_id_mapping_model'])
+        state_id_mapping_model = state_id_mapping_model_tuples[0][1]
+        if state_id_mapping_model:
+            if (
+                    len(state_id_mapping_model.state_names_to_ids) !=
+                    len(item.states)):
+                cls.errors['exploration state check'] = (
+                    'Model id %s: Corresponding StateIdMappingModel %s has '
+                    '%d states but model has %d' % (
+                        item.id, state_id_mapping_model.id,
+                        len(state_id_mapping_model.state_names_to_ids),
+                        len(item.states)))
+            for state_name in (
+                    state_id_mapping_model.state_names_to_ids.iterkeys()):
+                if state_name not in item.states:
+                    cls.errors['exploration state check'] = (
+                        'Model id %s: Corresponding StateIdMappingModel %s has '
+                        'state name %s but model doesn\'t' %
+                        (item.id, state_id_mapping_model.id, state_name))
 
-        cls._validate_external_id_relationships(item)
+    @classmethod
+    def _get_validation_functions(cls):
+        return [cls._validate_state_name]
+
+
+MODEL_TO_VALIDATOR_MAPPING = {
+    user_models.UserSubscriptionsModel: UserSubscriptionsModelValidator,
```

Comment:
Perhaps better for the key to just be the model class name (a string), otherwise this is bulkier than it needs to be.

Reply:
thanks!

Reply:
We have to list the Model classes for entity_classes_to_map_over anyways, so this probably isn't adding more